src/middlewared/middlewared/plugins/apps/ix_apps/query.py

```diff
@@ -0,0 +1,164 @@
+import os
+from dataclasses import dataclass
+from pkg_resources import parse_version
+
+from .docker.query import list_resources_by_project
+from .metadata import get_collective_metadata
+from .path import get_app_parent_config_path
+from .utils import PROJECT_PREFIX
+
+
+COMPOSE_SERVICE_KEY: str = 'com.docker.compose.service'
+
+
+@dataclass(frozen=True, eq=True)
+class VolumeMount:
+    source: str
+    destination: str
+    mode: str
+    type: str
+
+    def __hash__(self):
+        return hash((self.source, self.destination, self.type))
+
+
+def upgrade_available_for_app(
+    version_mapping: dict[str, dict[str, dict[str, str]]], app_metadata: dict
+) -> bool:
+    if version_mapping.get(app_metadata['train'], {}).get(app_metadata['name']):
+        return parse_version(app_metadata['version']) < parse_version(
+            version_mapping[app_metadata['train']][app_metadata['name']]['version']
+        )
+    else:
+        return False
+
+
+def list_apps(
+    train_to_apps_version_mapping: dict[str, dict[str, dict[str, str]]],
+    specific_app: str | None = None
+) -> list[dict]:
+    apps = []
+    app_names = set()
+    metadata = get_collective_metadata()
+    # This will only give us apps which are running or in deploying state
+    for app_name, app_resources in list_resources_by_project(
+        project_name=f'{PROJECT_PREFIX}{specific_app}' if specific_app else None,
+    ).items():
+        app_name = app_name[len(PROJECT_PREFIX):]
+        app_names.add(app_name)
+        if app_name not in metadata:
+            # The app is malformed or something is seriously wrong with it
+            continue
+
+        workloads = translate_resources_to_desired_workflow(app_resources)
+        # TODO: So when we stop an app, we remove all it's related resources and we wouldn't be in this for loop at all
+        #  however, when we stop docker service and start it again - the containers can be in exited state which means
+        #  we need to account for this.
+        #  This TODO however is for figuring out why app.start doesn't work with the compose actions we have in place
+        #  atm and should then we be maybe doing docker compose down on apps when stopping docker service
+        state = 'STOPPED'
```

Comment:
If state indeed remains `STOPPED` - probably better to give the default active workloads dict because it is not consistent with those apps which are actually stopped manually when they are queried?